### PR TITLE
fix deprecated `torch.cuda.amp.GradScaler` FutureWarning for pytorch 2.4+

### PR DESCRIPTION
**Pull request by Mon-ius**
With the release of PyTorch 2.4, the `torch.cuda.amp.GradScaler` API has been officially marked as deprecated.

This pull request (PR) addresses the following issue:

```py
accelerate/accelerator.py:494: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.scaler = torch.cuda.amp.GradScaler(**kwargs)
```

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for working on this deprecation. Could you please run `make quality` and `make style` on your PR?

**Comment by Mon-ius** [comment]
@BenjaminBossan I have finished ur requests, would u plz trigger the workflows 🤗

**Comment by Mon-ius** [comment]
strange that,  I run `make quality` okay on my side.

```sh
make quality
ruff check .
All checks passed!
ruff format --check .
150 files already formatted
doc-builder style src/accelerate docs/source --max_len 119 --check_only
```

**Comment by BenjaminBossan** [comment]
It's the doc-builder that is complaining. This is the diff I get from running `doc-builder style src/accelerate docs/source --max_len 119`:

```diff
@@ -127,10 +127,11 @@ def find_executable_batch_size(function: callable = None, starting_batch_size: i
 
 
     >>> @find_executable_batch_size(starting_batch_size=128)
-    ... def train(batch_size, model, optimizer): ...
+    ... def train(batch_size, model, optimizer):
+    ...     ...
 
 
-    ... train(model, optimizer)
+    >>> train(model, optimizer)
     ```
     """
     if function is None
```

**Comment by Mon-ius** [comment]
> ```diff
> @find_executable_batch_size(starting_batch_size=128)
> ```

Insteresting, but I didn't touch that file under `src/accelerate/utils/memory.py`, I run your command from my side, `doc-builder style src/accelerate docs/source --max_len 119`, nothing complaining.

I just update the code according to your side message. I am aware the `doc-builder` on `python3.8` is slight different from `python3.12` running on my side.

Could u plz trigger the CI again?

**Comment by BenjaminBossan** [comment]
Thanks @Mon-ius. I think one of your previous commits made a change to `memory.py` due to ruff or so. Thus let's completely undo any changes to the file and it should pass. That means, restoring the one line that's currently removed.

**Comment by Mon-ius** [comment]
@BenjaminBossan I just deleted the file and download the original one, seems just one blank line make the CI failed.

Could you plz trigger it again? 🤗

**Comment by Mon-ius** [comment]
Hi, @BenjaminBossan

Seems all checks have been passed now, could u plz to merge it 🤗

**Review by BenjaminBossan — COMMENTED** [review]
I checked this change with older PyTorch versions and unfortunately, it is only supported starting from 2.3 on. Since we want to support older versions as well, we need to make a torch version check.

Probably a similar argument applies to the other  grad checkers, like `torch.npu.amp.GradScaler`, but I haven't tested.

**Comment by Mon-ius** [comment]
@BenjaminBossan  Thx for your review, I applied a simple pytorch version check for 2.3 on GPU device, could u plz to trigger the CI?

**Comment by Mon-ius** [comment]
For npu device, I currently dont hold one to test, but I also applied the strategy on last seperate commit 🤗

File: src/accelerate/checkpointing.py [hunk 1]
```diff
@@ -86,8 +86,8 @@ def save_accelerator_state(
             The current process index in the Accelerator state
         step (`int`):
             The current step in the internal step tracker
-        scaler (`torch.cuda.amp.GradScaler`, *optional*):
-            An optional gradient scaler instance to save
+        scaler (`torch.amp.GradScaler`, *optional*) for pytorch>2.3:
+            An optional gradient scaler instance to save; for lower version, check `torch.cuda.amp.GradScaler`
```
Comment (BenjaminBossan) [inline src/accelerate/checkpointing.py:90]: I think we don't need this extra explanation in the docstring, it should be quite clear as is what this refers to.

File: src/accelerate/utils/dataclasses.py [hunk 2]
```diff
@@ -209,8 +209,9 @@ def register_comm_hook(self, model):
 class GradScalerKwargs(KwargsHandler):
     """
     Use this object in your [`Accelerator`] to customize the behavior of mixed precision, specifically how the
-    `torch.cuda.amp.GradScaler` used is created. Please refer to the documentation of this
-    [scaler](https://pytorch.org/docs/stable/amp.html?highlight=gradscaler) for more information on each argument.
+    `torch.amp.GradScaler` used is created for pytoch>2.3 or `torch.cuda.amp.GradScaler` for lower version. Please
```
Comment (BenjaminBossan) [inline src/accelerate/utils/dataclasses.py:212]: Same, let's not overexplain.

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -494,11 +495,17 @@ def __init__(
             elif is_musa_available():
                 self.scalar = torch.musa.amp.GradScaler(**kwargs)
             elif is_npu_available():
-                self.scaler = torch.npu.amp.GradScaler(**kwargs)
+                if version.parse(torch.__version__) > version.parse("2.3"):
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:498]: Since we both can't test this, let's not touch NPU for now.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the version fixes. I have a few smaller comments.

It would probably make sense to factor out the grad scaler instantiation to a utility function to avoid code duplication, but I'll let Zach be the judge of that once he's back in office. For now, it can stay as is.

**Comment by Mon-ius** [comment]
@BenjaminBossan Thanks again for the comments, I have reverted the code related to the documentation and NPU. 🤗

**Comment by Mon-ius** [comment]
seems there is an CI self issue, which lacks `pillow` on test environment for [failed](https://github.com/huggingface/accelerate/actions/runs/11145916259/job/30979916686?pr=3132)

```sh
=========================== short test summary info ============================
FAILED tests/trainer/test_trainer.py::TrainerIntegrationTest::test_trainer_saves_image_processor - ImportError: 
CLIPImageProcessor requires the PIL library but it was not found in your environment. You can install it with pip:
`pip install pillow`. Please note that you may need to restart your runtime after installation.
FAILED tests/trainer/test_trainer.py::TrainerIntegrationTest::test_trainer_saves_processor - ImportError: 
CLIPImageProcessor requires the PIL library but it was not found in your environment. You can install it with pip:
`pip install pillow`. Please note that you may need to restart your runtime after installation.
=========== 2 failed, 172 passed, 81 skipped, 42 warnings in 28.79s ============
```

**Comment by BenjaminBossan** [comment]
Yes, appears to be unrelated.

**Comment by Mon-ius** [comment]
> Thanks for the version fixes. I have a few smaller comments.
> 
> It would probably make sense to factor out the grad scaler instantiation to a utility function to avoid code duplication, but I'll let Zach be the judge of that once he's back in office. For now, it can stay as is.

@BenjaminBossan, how about Zach's review of this PR 🤗

**Review by muellerzr — APPROVED** [review]
Thanks, this looks okay to me. This hints that we should probably have a `get_grad_scaler` util func in Accelerate which can handle the backends automatically as part of the ops util package, but for the scope of this PR this works great. 

Thanks!

**Pull request merged**